.changeset/poor-berries-occur.md

```diff
@@ -0,0 +1,51 @@
+---
+"astro": minor
+---
+
+Adds experimental support for the `astro:env` API. It lets you define a type-safe schema for your environments variables, and where they should be available (on the server or the client).
+
+Start by declaring a schema in your Astro config in `experimental.env.schema`. Use the `envField` help to make your life easier:
+
+```js
+// astro.config.mjs
+import { defineConfig, envField } from "astro/config"
+
+export default defineConfig({
+    experimental: {
+        env: {
+            schema: {
+                PUBLIC_API_URL: envField.string({ context: "client", access: "public", optional: true }),
+                PUBLIC_PORT: envField.number({ context: "server", access: "public", default: 4321 }),
+                API_SECRET: envField.string({ context: "server", access: "secret" }),
```

Comment:
Re-reading this, since `PUBLIC_` already hinted what the `access` would be, maybe we should remove `access` altogether from the expose API, and we can auto set it internally?

Reply:
I quite like forcing people to be super explicit but I don't mind inferring it as well. I'd like other opinions before refactoring

Reply:
I would also prefer expressiveness over "magic"

Reply:
Alright then. Let's see how this goes.